posts/shadow-dom-and-event-propagation.md

```diff
@@ -0,0 +1,267 @@
+
+```js
+const div = document.createElement("div");
+const shadowRoot = div.attachShadow({ mode: "open" });
+
+shadowRoot.addEventListener("test", (evt) => {
+  console.log(">> Shadow root listener invoked", evt);
+});
+
+const evt = new Event("test");
+shadowRoot.dispatchEvent(evt);
+```
+
+The shadow root is the root node for a shadow tree. It is possible to listen to any event originating from a shadow tree by adding the appropriate event listener directly on the shadow root.
+
+Similar to how events work in the light DOM, events only bubble in the shadow DOM if the [`bubbles`](https://developer.mozilla.org/en-US/docs/Web/API/Event/bubbles) option is set to `true`: `new Event('test', { bubbles: true })`. Events dispatch from within the shadow tree invokes the shadow root listeners if the event is marked as `bubbles`.
+
+## Escaping the shadow trees using `composed`
+
+By default, events don't propagate outside shadow trees. This default behavior ensures that internal DOM events don't inadvertently leak into the document.
+
+In the example, dispatching a bubbling event from the `span#d` propagates to `p#c` and the shadow root, but it stops there. The event never reaches `div#a` since the element lives outside the shadow tree.
+
+<event-visualizer label="Single shadow tree" event-bubbles>
+  <template>
+    <div id="a">
+      <template shadowroot="open">
+        <p id="c">
+          <span id="d" target></span>
+        </p>
+        <p id="e"></p>
+      </template>
+      <img id="b"></img>
+    </div>
+  </template>
+</event-visualizer>
+
+**For an event to traverse shadow DOM boundaries, it has to be configured as a [`composed`](https://developer.mozilla.org/en-US/docs/Web/API/Event/composed)**: `new Event('test', { bubbles: true, composed: true })`. Going back to the previous example, if `composed` is enabled, you can see the event escaping the shadow tree and reaching to `div#a`.
+
+When the `composed` option is set to `true`, the dispatch event not only traverses its shadow boundary but also any other parent boundary. In the following example, the bubbling and composed event dispatch from `span#e` bubble to `div#c`, `div#a`, and their respective shadow roots.
+
+<event-visualizer label="Nested shadow trees" event-bubbles event-composed>
+  <template>
+    <div id="a">
+      <template shadowroot="open">
+        <div id="c">
+          <template shadowroot="open">
+            <span id="e" target></span>
+          </template>
+          <p id="d"></p>
+        </div>
+      </template>
+      <img id="b"></img>
+    </div>
+  </template>
+</event-visualizer>
+
+This might be counterintuitive but, a composed event always propagates outside the shadow boundary regardless if it is bubbling or not. Give a shot and set `bubbles` to false in the previous example to see how the event propagates.
```

Comment:
```suggestion
This might be counterintuitive, but a composed event always propagates outside the shadow boundary regardless of whether it is bubbling or not. Give it a shot and set `bubbles` to false in the previous example to see how the event propagates.
```